File: tests/test_op/test_dynamic_import_op/AdditionOP/AdditionOP.cpp
/*!
 * \file AdditionOP.cpp
 * \brief a C++ addition operator
 */

// include Mobula OP head file
#include "mobula_op.h"
using namespace mobula;

/**
 * \brief the forward kernel implementation of AdditionOP (out_c = in_a + in_b)
 * \param n      the number of elements
 * \param a      input array
 * \param b      input array
 * \param c     output array
 */
// use `MOBULA_KERNEL` macro to declare a kernel function
// the kernel function will be adapted for CPU and GPU
MOBULA_KERNEL addition_op_forward_kernel(const int n, const float* a,
                                         const float* b, float* c) {
  // use parallel for-loop
  // `parfor(number-of-iterations, function)`
  // please NOTE Thread Safety in `parfor`
  parfor(n, [&](int i) { c[i] = a[i] + b[i]; });
}

File: tests/test_op/test_dynamic_import_op/AdditionOP/AdditionOP.py
import mobula


@mobula.op.register
class AdditionOP:
    def forward(self, a, b):
        c = self.y
        mobula.func.addition_op_forward(a.size, a, b, c)

    def backward(self, dc):
        return [dc, dc]

    def infer_shape(self, in_shape):
        assert list(in_shape[0]) == list(in_shape[1])
        return in_shape, [in_shape[0]]
